### PR TITLE
New rule: Argument variable must match type

### PR DESCRIPTION
**Pull request by alamirault**
Rule to ensure consistency :smile: 

https://github.com/symfony/symfony-docs/pull/17687#issuecomment-1376119707

I think this rule must be configurable, next rules names will be easy !

- [x] Documentation

- [x] Adjustments

### CODE REVIEW

File: src/Rule/ArgumentTypeMatchName.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace App\Rule;
+
+use App\Value\Lines;
+use App\Value\RuleGroup;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ArgumentTypeMatchName extends AbstractRule implements LineContentRule, Configurable
```
Comment (alamirault) [inline src/Rule/ArgumentTypeMatchName.php:20]: Have you a better name ?

`ArgumentNameMustMatchType` ?

File: src/Rule/ArgumentTypeMatchName.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            // This regex match argument type with bad argument name
+            $regex = sprintf(
+                '/%s \$(?!%s)/',
+                $argument['type'],
+                $argument['name']
+            );
+
+            if ($line->match($regex)) {
+                $messageParts[] = sprintf(
+                    'Please name the argument "%s $%s"',
```
Comment (alamirault) [inline src/Rule/ArgumentTypeMatchName.php:66]: I can probably change to `Please rename $configurator to $containerConfigurator`
With this regex, but is it better for user ? And for maintenance
`ContainerConfigurator (?!\$containerConfigurator)(?<actualName>[a-z-A-Z\$]+)`

File: src/Rule/ArgumentTypeMatchName.php (hunk 1 above)
Comment (OskarStark) [reply src/Rule/ArgumentTypeMatchName.php:20]: ArgumentVariableMustMatchType ?

File: src/Rule/ArgumentTypeMatchName.php (hunk 2 above)
Comment (OskarStark) [reply src/Rule/ArgumentTypeMatchName.php:66]: I like it

**Review by OskarStark — APPROVED** [review]

File: src/Rule/ArgumentTypeMatchName.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    public static function getGroups(): array
+    {
+        return [RuleGroup::Symfony()];
+    }
+
+    public function check(Lines $lines, int $number): ?string
+    {
+        $lines->seek($number);
+        $line = $lines->current()->clean();
```
Comment (OskarStark) [inline src/Rule/ArgumentTypeMatchName.php:53]: I think we only need to do the checks if we are inside a PHP code block. Check some other rules how I achieved that

**Comment by OskarStark** [comment]
Not sure you know, but there is a command to generate the documentation :-)

File: src/Rule/ArgumentVariableMustMatchType.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if (!RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP)
+            && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_ANNOTATIONS)
+            && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_ATTRIBUTES)
+        ) {
+            return null;
+        }
+
+        $lines->next();
+        $lines->next();
+
```
Comment (alamirault) [inline src/Rule/ArgumentVariableMustMatchType.php:72]: I saw that on other rules is it the right way ?

**Comment by alamirault** [comment]
> Not sure you know, but there is a command to generate the documentation :-)

I didn't know, it's great !

However nested OptionResolver is not supported and not easy to adapt for better ouput in this case

File: src/Rule/ArgumentVariableMustMatchType.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+use App\Value\Lines;
+use App\Value\RuleGroup;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @Description("Make sure argument variable name match for type")
+ */
+class ArgumentVariableMustMatchType extends AbstractRule implements LineContentRule, Configurable
+{
+    /** @var array<array{type: string, name:string}> */
```
Comment (OskarStark) [inline src/Rule/ArgumentVariableMustMatchType.php:27]:
```suggestion
    /** @var array<array{type: string, name: string}> */
```

**Review by OskarStark — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Comment by OskarStark** [comment]
Thank you!

**Comment by dkarlovi** [comment]
Please note this as a possible collision https://github.com/symfony/symfony-docs/issues/17795

**Pull request merged**